File: internal/review/parser_test.go
package review_test

import (
	"testing"

	"github.com/CrowdStrike/codestrike/internal/review"
)

func TestParsePRURL(t *testing.T) {
	tests := []struct {
		name    string
		url     string
		want    review.PRReference
		wantErr bool
	}{
		{
			name: "standard github URL",
			url:  "https://github.com/CrowdStrike/codestrike/pull/42",
			want: review.PRReference{Provider: review.ProviderGitHub, Owner: "CrowdStrike", Repo: "codestrike", Number: 42},
		},
		{
			name: "github trailing slash",
			url:  "https://github.com/owner/repo/pull/7/",
			want: review.PRReference{Provider: review.ProviderGitHub, Owner: "owner", Repo: "repo", Number: 7},
		},
		{
			name: "standard bitbucket URL",
			url:  "https://bitbucket.org/workspace/my-repo/pull-requests/123",
			want: review.PRReference{Provider: review.ProviderBitbucket, Owner: "workspace", Repo: "my-repo", Number: 123},
		},
		{
			name: "bitbucket trailing slash",
			url:  "https://bitbucket.org/workspace/repo/pull-requests/5/",
			want: review.PRReference{Provider: review.ProviderBitbucket, Owner: "workspace", Repo: "repo", Number: 5},
		},
		{
			name:    "invalid URL missing pull segment",
			url:     "https://github.com/owner/repo/issues/1",
			wantErr: true,
		},
		{
			name:    "invalid PR number",
			url:     "https://github.com/owner/repo/pull/abc",
			wantErr: true,
		},
		{
			name:    "too short URL",
			url:     "https://github.com/pull/1",
			wantErr: true,
		},
		{
			name:    "invalid bitbucket PR number",
			url:     "https://bitbucket.org/workspace/repo/pull-requests/abc",
			wantErr: true,
		},
		{
			name:    "too short bitbucket URL",
			url:     "https://bitbucket.org/pull-requests/1",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := review.ParsePrUrl(tt.url)
			if tt.wantErr {
				if err == nil {
					t.Fatal("expected error")
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got != tt.want {
				t.Errorf("got %+v, want %+v", got, tt.want)
			}
		})
	}
}

File: tests/review/parser.test.ts
import { describe, it, expect } from 'vitest';
import { parsePrUrl, PROVIDER_GITHUB, PROVIDER_BITBUCKET } from '../../src/review/parser.js';

describe('parsePrUrl', () => {
  it('parses a GitHub PR URL', () => {
    const ref = parsePrUrl('https://github.com/CrowdStrike/codestrike/pull/42');
    expect(ref.provider).toBe(PROVIDER_GITHUB);
    expect(ref.owner).toBe('CrowdStrike');
    expect(ref.repo).toBe('codestrike');
    expect(ref.number).toBe(42);
  });

  it('parses a GitHub PR URL with trailing slash', () => {
    const ref = parsePrUrl('https://github.com/CrowdStrike/codestrike/pull/42/');
    expect(ref.provider).toBe(PROVIDER_GITHUB);
    expect(ref.number).toBe(42);
  });

  it('parses a Bitbucket PR URL', () => {
    const ref = parsePrUrl('https://bitbucket.org/myworkspace/myrepo/pull-requests/7');
    expect(ref.provider).toBe(PROVIDER_BITBUCKET);
    expect(ref.owner).toBe('myworkspace');
    expect(ref.repo).toBe('myrepo');
    expect(ref.number).toBe(7);
  });

  it('throws on missing /pull/ segment', () => {
    expect(() => parsePrUrl('https://github.com/CrowdStrike/codestrike')).toThrow(
      'missing /pull/ or /pull-requests/ segment',
    );
  });

  it('throws on non-numeric PR number', () => {
    expect(() => parsePrUrl('https://github.com/CrowdStrike/codestrike/pull/abc')).toThrow('invalid PR number');
  });

  it('throws on URL with insufficient segments', () => {
    expect(() => parsePrUrl('https://pull/42')).toThrow('missing owner or repo');
  });
});
